torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -282,62 +287,128 @@ InputsIdLookup::IdLookupReturn InputsIdLookup::lookupId(
   return ret;
 }
 
-FusionExecutorCache::FusionExecutorCache(std::unique_ptr<Fusion>&& fusion)
-    : fusion_(std::move(fusion)) {
-  FUSER_PERF_SCOPE("FusionExecutorCache::FusionExecutorCache");
+OptimizedFusionGraph::OptimizedFusionGraph(
+    Fusion* fusion,
+    const at::ArrayRef<IValue>& inputs) {
+  FUSER_PERF_SCOPE("OptimizedFusionGraph::OptimizedFusionGraph");
+
+  // Make a copy of fusion and do translation on the copied graph
+  auto transformed_fusion = std::make_unique<Fusion>(*fusion);
+
+  // Run segmentation on the copied fusion
+  SchedulerRuntimeInfo runtime_info(transformed_fusion.get(), inputs, true);
+
+  // This is where pre-segment passes such as translateWelford will go
 
   //! Try to schedule the complete fusion
   const auto maybe_complete_fusion_scheduler =
-      SchedulerEntry::proposeHeuristics(fusion_.get());
+      SchedulerEntry::proposeHeuristics(transformed_fusion.get(), runtime_info);
 
   //! Decide if this fusion is segmented or not
   const bool segmented = !maybe_complete_fusion_scheduler.has_value();
 
   if (segmented) {
-    // Segment the fusion through FusionSegmenter and
+    // Take ownership and segment transformed fusion
+    fusion_segments_ =
+        SegmentCandidateFinder::segment(std::move(transformed_fusion), inputs);
+
     //  initialize the caching for segmented heuristics
-    fusion_segments_ = fusion_->segment();
     fusion_kernel_runtime_cache_.initSegmentCache(fusion_segments_.get());
     if (isDebugDumpEnabled(DebugDumpOption::FusionSegments)) {
       fusion_segments_->print();
     }
   } else {
+    // Take ownership of the transformed fusion
+    single_kernel_fusion_ = std::move(transformed_fusion);
     // Initialize single kernel case
     fusion_kernel_runtime_cache_.initSingleKernelCache(
-        fusion_.get(), maybe_complete_fusion_scheduler.value());
+        single_kernel_fusion_.get(), maybe_complete_fusion_scheduler.value());
     // In the case that the fusion isn't segmented but user
     //  wants segmented fusion in the debug print. Will
     //  print math of the composite fusion as placeholder
     if (isDebugDumpEnabled(DebugDumpOption::FusionSegments)) {
-      fusion_->printMath();
+      single_kernel_fusion_->printMath();
     }
   }
 }
 
+bool OptimizedFusionGraph::canReUseFor(const at::ArrayRef<IValue>& inputs) {
+  // Need to construct the runtime info here to ensure correct evaluator
+  // binding, each optimized fusion graph owns a different (potentially
+  // translated) copy
+
+  SchedulerRuntimeInfo runtime_info(fusion(), inputs);
+  // TODO: welford translation checks go here
+  return fusion_kernel_runtime_cache_.canReUseFor(runtime_info);
+}
+
+Fusion* OptimizedFusionGraph::fusion() {
```

Comment:
So this returns a scheduled fusion for single_kernel_fusion, but the original complete fusion for fusion_segments_?
nitpick: Maybe put that in a comment to avoid confusion.

Reply:
This is actually un-scheduled original fusion for both cases. Added comments to clarify. Thanks for pointing this out.